File: Java/src/_0879_Profitable_Schemes/Solution.java
package _0879_Profitable_Schemes;

/**
 * Date: 2021/6/9 10:51
 * 执行用时：28 ms, 在所有 Java 提交中击败了94.74%的用户
 * 内存消耗：37.7 MB, 在所有 Java 提交中击败了64.84%的用户
 */
public class Solution {

    static final int mod = (int) (1e9 + 7);

    public int profitableSchemes(int n, int m, int[] group, int[] profit) {
        int[][] f = new int[n + 1][m + 1];
        for (int i = 0; i <= n; i++) f[i][0] = 1;
        for (int i = 0; i < group.length; i++) {
            int g = group[i], p = profit[i];
            for (int j = n; j >= g; j--)
                for (int k = m; k >= 0; k--)
                    f[j][k] = (f[j][k] + f[j - g][Math.max(0, k - p)]) % mod;
        }
        return f[n][m];
    }

    public static void main(String[] args) {

        int[] group = {2, 2}, profit = {2, 3};
        System.out.println((new Solution()).profitableSchemes(5, 3, group, profit));  // 2
    }
}

File: Python/_0879_Profitable_Schemes/Solution.py
# coding=utf-8
# Date: 2021/6/9 11:08
from typing import List


# 执行用时：1584 ms, 在所有 Python3 提交中击败了59.54%的用户
# 内存消耗：15.1 MB, 在所有 Python3 提交中击败了65.13%的用户
class Solution:
    def profitableSchemes(self, n: int, m: int, group: List[int], profit: List[int]) -> int:
        mod = (int)(1e9 + 7)
        f = [[0 for _ in range(m + 1)] for _ in range(n + 1)]
        for i in range(n + 1):
            f[i][0] = 1
        for i in range(len(group)):
            g = group[i]
            p = profit[i]
            for j in range(n, g - 1, -1):
                for k in range(m, -1, -1):
                    f[j][k] = (f[j][k] + f[j - g][max(0, k - p)]) % mod
        return f[n][m]


if __name__ == "__main__":
    print(Solution().profitableSchemes(5, 3, [2, 2], [2, 3]))  # 2
